src/rpc/format/cbor/convert.rs

```diff
@@ -0,0 +1,182 @@
+								Ok(Value::from(Thing::from((tb, Id::from(i128::from(id) as i64)))))
+							}
+							(Data::Text(tb), Data::Array(id)) => Ok(Value::from(Thing::from((
+								tb,
+								Id::from(
+									id.into_iter()
+										.map(|v| Value::try_from(Cbor(v)))
+										.collect::<Result<Vec<Value>, &str>>()?,
+								),
+							)))),
+							(Data::Text(tb), Data::Map(id)) => Ok(Value::from(Thing::from((
+								tb,
+								Id::from(
+									id.into_iter()
+										.map(|(k, v)| {
+											let k =
+												Value::try_from(Cbor(k)).map(|k| k.as_raw_string());
+											let v = Value::try_from(Cbor(v));
+											Ok((k?, v?))
+										})
+										.collect::<Result<BTreeMap<String, Value>, &str>>()?,
+								),
+							)))),
+							_ => Err("Expected a CBOR array with 2 elements, a text data type, and a valid ID type"),
+						},
+						_ => Err("Expected a CBOR text data type, or a CBOR array with 2 elements"),
+					},
+					// An unknown tag
+					_ => Err("Encountered an unknown CBOR tag"),
+				}
+			}
+			_ => Err("Encountered an unknown CBOR data type"),
+		}
+	}
+}
+
+impl TryFrom<Value> for Cbor {
+	type Error = &'static str;
+	fn try_from(val: Value) -> Result<Self, &'static str> {
+		match val {
+			Value::None => Ok(Cbor(Data::Tag(TAG_NONE, Box::new(Data::Null)))),
+			Value::Null => Ok(Cbor(Data::Null)),
+			Value::Bool(v) => Ok(Cbor(Data::Bool(v))),
+			Value::Number(v) => match v {
+				Number::Int(v) => Ok(Cbor(Data::Integer(v.into()))),
+				Number::Float(v) => Ok(Cbor(Data::Float(v))),
+				Number::Decimal(v) => {
+					Ok(Cbor(Data::Tag(TAG_DECIMAL, Box::new(Data::Text(v.to_string())))))
+				}
+			},
+			Value::Strand(v) => Ok(Cbor(Data::Text(v.0))),
+			Value::Duration(v) => {
+				Ok(Cbor(Data::Tag(TAG_DURATION, Box::new(Data::Text(v.to_raw())))))
+			}
+			Value::Datetime(v) => {
+				Ok(Cbor(Data::Tag(TAG_DATETIME, Box::new(Data::Text(v.to_raw())))))
+			}
+			Value::Uuid(v) => Ok(Cbor(Data::Tag(TAG_UUID, Box::new(Data::Text(v.to_raw()))))),
+			Value::Array(v) => Ok(Cbor(Data::Array(
+				v.into_iter()
+					.map(|v| {
+						let v = Cbor::try_from(v)?.0;
+						Ok(v)
+					})
+					.collect::<Result<Vec<Data>, &str>>()?,
+			))),
+			Value::Object(v) => Ok(Cbor(Data::Map(
+				v.into_iter()
+					.map(|(k, v)| {
+						let k = Data::Text(k);
+						let v = Cbor::try_from(v)?.0;
+						Ok((k, v))
+					})
+					.collect::<Result<Vec<(Data, Data)>, &str>>()?,
+			))),
+			Value::Bytes(v) => Ok(Cbor(Data::Bytes(v.into_inner()))),
+			Value::Thing(v) => Ok(Cbor(Data::Tag(
+				TAG_RECORDID,
+				Box::new(Data::Array(vec![
+					Data::Text(v.tb),
+					match v.id {
+						Id::Number(v) => Data::Integer(v.into()),
+						Id::String(v) => Data::Text(v),
+						Id::Array(v) => Cbor::try_from(Value::from(v))?.0,
+						Id::Object(v) => Cbor::try_from(Value::from(v))?.0,
+						Id::Generate(_) => unreachable!(),
+					},
+				])),
+			))),
+			// We shouldn't reach here
+			_ => Ok(Cbor(Data::Null)),
```

Comment:
I think it would be favorible to throw an error here

Reply:
I agree @kearfy. If it shouldn't possible to reach here then `unreachable!()` will do too. Returning an error or panicking will make it easier to know if we do manage to reach here due to a bug.

```suggestion
			_ => unreachable!(),
```